framework/src/base/NonlinearEigenSystem.C

```diff
@@ -42,14 +43,48 @@ assemble_matrix(EquationSystems & es, const std::string & system_name)
   p->computeJacobian(
       *eigen_system.current_local_solution.get(), *eigen_system.matrix_A, Moose::KT_NONEIGEN);
 
+  Mat petsc_mat_A = static_cast<PetscMatrix<Number> &>(*eigen_system.matrix_A).mat();
+
+  PetscObjectComposeFunction((PetscObject)petsc_mat_A,
+                             "formJacobian",
+                             Moose::SlepcSupport::mooseSlepcEigenFormJacobianA);
+  PetscObjectComposeFunction((PetscObject)petsc_mat_A,
+                             "formFunction",
+                             Moose::SlepcSupport::mooseSlepcEigenFormFunctionA);
+
+  PetscContainer container;
+  PetscContainerCreate(eigen_system.comm().get(), &container);
+  PetscContainerSetPointer(container, p);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formJacobianCtx", NULL);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formJacobianCtx", (PetscObject)container);
```

Comment:
Yes. The first line detaches the old function if we have, and the second line attaches the  new function